packages/web-client/app/models/workflow/workflow-card.ts

```diff
@@ -19,44 +19,60 @@ type FailureCheckResult = {
 export type CheckResult = SuccessCheckResult | FailureCheckResult;
 
 interface WorkflowCardOptions {
+  cardName?: string;
   author: Participant;
   componentName: string; // this should eventually become a card reference
   includeIf(this: WorkflowCard): boolean;
   check(this: WorkflowCard): Promise<CheckResult>;
 }
 
 export class WorkflowCard extends WorkflowPostable {
+  cardName: string;
   componentName: string;
-  check?: (this: WorkflowCard) => Promise<CheckResult>;
+  check: (this: WorkflowCard) => Promise<CheckResult> = () => {
+    return Promise.resolve({ success: true });
+  };
 
   constructor(options: Partial<WorkflowCardOptions>) {
     super(options.author!, options.includeIf);
     this.componentName = options.componentName!;
+    this.cardName = options.cardName || '';
+
     this.reset = () => {
       if (this.isComplete) {
         this.isComplete = false;
       }
     };
-    this.check = options.check;
+    if (options.check) {
+      this.check = options.check;
+    }
   }
   get session(): WorkflowSession | undefined {
     return this.workflow?.session;
   }
 
-  @action onComplete() {
-    if (this.check) {
-      this.check().then((checkResult) => {
-        if (checkResult.success) {
-          // visible-postables-will-change starts test waiters in animated-workflow.ts
-          this.workflow?.emit('visible-postables-will-change');
-          this.isComplete = true;
-        } else {
-          this.workflow?.cancel(checkResult.reason);
-        }
-      });
-    } else {
+  @action async onComplete() {
+    if (this.isComplete) return;
+    let checkResult = await this.check();
+    if (checkResult.success) {
+      // visible-postables-will-change starts test waiters in animated-workflow.ts
       this.workflow?.emit('visible-postables-will-change');
       this.isComplete = true;
+    } else {
+      this.workflow?.cancel(checkResult.reason);
+    }
+
+    if (this.isComplete && this.cardName) {
+      const existingCompletedCardNames =
+        this.session?.state.completedCardNames || [];
+
+      if (!existingCompletedCardNames.includes(this.cardName)) {
+        this.session?.updateMany({
+          completedCardNames: [...existingCompletedCardNames, this.cardName],
+          completedMilestonesCount: this.workflow?.completedMilestoneCount,
+          milestonesCount: this.workflow?.milestones.length,
```

Comment:
I can't find where we're using `completedMilestonesCount` and `milestonesCount`, what are we using them for?

Reply:
I added them for convenience - to support the progress display on unfinished workflows for when we add this list (check the design below). Do you think it would be more suitable to calculate progress by looking at the finished cards instead?

![image](https://user-images.githubusercontent.com/273660/133067577-a86f44e6-2330-4b87-b030-01d7d4aa64d5.png)

Reply:
Ah I see, thanks! I'm not sure what the best way to do it is, in an ideal world I would like to have the milestones/completion percentage be derived state

Reply:
Sounds good. We'll evaluate then - I'll leave this in for now.